### PR TITLE
DHT Benchmark with asynchronous w/r

### PR DESCRIPTION
**Pull request by MuXauJl11110**
Based on #350

### CODE REVIEW

**Comment by justheuristic** [comment]
Hi!
Please refer to [contributing.md#code-style](https://github.com/learning-at-home/hivemind/blob/master/CONTRIBUTING.md#code-style) for code style

File: benchmarks/benchmark_dht.py [hunk 1]
```diff
@@ -1,129 +1,164 @@
...
+
+
+async def get_task(peer, key):
+    latest = bool(random.getrandbits(1))
+    value, expiration = await peer.get(key, latest=latest, return_future=True)
+
+    return value, expiration
+
+
+async def corouting_task(
```
Comment (justheuristic) [inline benchmarks/benchmark_dht.py:35]: Consider renaming this into something more informative, e.g. store_get_task

File: benchmarks/benchmark_dht.py [hunk 2]
```diff
@@ -1,129 +1,164 @@
...
+    
+    store_ok = await peer.store(
+        key, subkey=subkey, value=value, 
+        expiration_time=hivemind.get_dht_time() + expiration, return_future=True
     )
 
+    return store_ok
+
+
+async def get_task(peer, key):
```
Comment (justheuristic) [inline benchmarks/benchmark_dht.py:28]: This is essentially a one-liner, perhaps it would be best to inline it into the main coro

**Comment by justheuristic** [comment]
Let's create a task that runs this benchmark (and the two remaining ones) on pull-request using the same python version as codecov_in_develop_mode

More on other benchmarks:
https://learning-at-home.readthedocs.io/en/latest/user/benchmarks.html


```
python benchmark_tensor_compression.py
python benchmark_throughput.py --preset minimalistic
[enter whichever command runs your benchmark here]
```

For instance, you can create a separate job in [run_tests.yml](https://github.com/learning-at-home/hivemind/blob/master/.github/workflows/run-tests.yml)

Rationale:
- **why CI**: we often break benchmarks, introducing CI will ensure that all benchmarks work
- **why separate CI job**: we could add this to tests, but it would extend the (already long) test runtime

**Comment by justheuristic** [comment]
Before merge
- add CI job with benchmarks
- measure global correctness rate and time to store/get (based on #350 )
- change latest from random variable into a parameter (based on #350 )
- TODO @justheuristic re-run benchmark for docs/bechmarking.html

If you still have time after that, lets implement the failure rate as described in #350

File: benchmarks/benchmark_dht.py [hunk 3]
```diff
@@ -1,129 +1,164 @@
...
-    expert_uids = list(
-        set(
-            f"expert.{random.randint(0, 999)}.{random.randint(0, 999)}.{random.randint(0, 999)}"
-            for _ in range(num_experts)
-        )
-    )
-    logger.info(f"Sampled {len(expert_uids)} unique ids (after deduplication)")
-    random.shuffle(expert_uids)
+    task_list = [
+        loop.create_task(
```
Comment (justheuristic) [inline benchmarks/benchmark_dht.py:131]: this is not recommended, better make the whole function asynchronous and use `asyncio.run`

File: benchmarks/benchmark_dht.py [hunk 4]
```diff
@@ -1,129 +1,164 @@
...
-                and expert.uid == expert_uids[start + i]
-                and expert.endpoint == endpoints[start // expert_batch_size]
-            ):
-                successful_gets += 1
-
-    if time.perf_counter() - benchmark_started > expiration:
-        logger.warning(
-            "keys expired midway during get requests. If that isn't desired, increase expiration_time param"
-        )
+    loop.run_until_complete(asyncio.wait(task_list))
```
Comment (justheuristic) [inline benchmarks/benchmark_dht.py:138]:
```suggestion
    loop.run_until_complete(asyncio.gather(*task_list))
```

File: .github/workflows/check-style.yml [hunk 5]
```diff
@@ -1,6 +1,6 @@
 name: Check style
 
-on: [ push ]
+on: [ push, pull_request ]
```
Comment (mryab) [inline .github/workflows/check-style.yml:3]: I'm not sure if these changes are necessary for this PR; if possible, I'd keep the diff as short as possible

Comment (justheuristic) [reply .github/workflows/check-style.yml:3]: They are, tests do not run in forks w/o this change

File: benchmarks/benchmark_dht.py [hunk 6]
```diff
@@ -42,88 +127,66 @@ def benchmark_dht(
...
+    parser.add_argument("--num_get_peers", type=int, default=8, required=False)
+    parser.add_argument("--wait_after_iteration", type=float, default=0, required=False)
+    parser.add_argument("--delay", type=float, default=0, required=False)
     parser.add_argument("--wait_timeout", type=float, default=5, required=False)
-    parser.add_argument("--random_seed", type=int, default=random.randint(1, 1000))
-    parser.add_argument("--increase_file_limit", action="store_true")
-    args = vars(parser.parse_args())
+    parser.add_argument("--expiration", type=float, default=300, required=False)
+    parser.add_argument("--latest", type=bool, default=True, required=False)
+    parser.add_argument("--failure_rate", type=float, default=0.1, required=False)
```
Comment (justheuristic) [inline benchmarks/benchmark_dht.py:188]: Let's keep the option to increase the file limit, for the sake of benchmarking with a very large number of peers.

File: benchmarks/benchmark_dht.py [hunk 7]
```diff
@@ -1,35 +1,120 @@
...
+    store_times = []
+    get_times = []
+
+    for _ in range(total_num_rounds):
+        key = uuid.uuid4().hex
+        value = uuid.uuid4().hex
+
+        store_start = time.perf_counter()
+        store_peers = random.sample(peers, min(num_store_peers, len(peers)))
+        store_tasks = [store_task(peer, key, value, expiration) for peer in store_peers]
```
Comment (justheuristic) [inline benchmarks/benchmark_dht.py:50]:
```suggestion
        subkeys = [uuid.uuid4().hex for peer in store_peers]
        store_tasks = [peer.store(
            peer, key, subkey=subkey, value=value, get_dht_time() + expiration, return_future=True)
            for peer, subkey in zip(store_peers, subkeys)]
```

To the best of my knowledge, this coro is only used once. I would recommend either of:
- inlining it: see suggestion above
- or formatting it: add docstring and type hints

File: benchmarks/benchmark_dht.py [hunk 8]
```diff
@@ -1,35 +1,120 @@
...
+    peers: list,
+    total_num_rounds: int,
+    num_store_peers: int,
+    num_get_peers: int,
+    wait_after_iteration: float,
+    delay: float,
+    expiration: float,
+    latest: bool,
+    failure_rate: float,
+):
```
Comment (justheuristic) [inline benchmarks/benchmark_dht.py:37]:
```suggestion
) -> Tuple[int, int, int, int]:
    """Iteratively choose random peers to store data onto the dht, then retreive with another random subset of peers"""
```

File: benchmarks/benchmark_dht.py [hunk 9]
```diff
@@ -1,35 +1,120 @@
...
+            successful_stores_per_iter += store_ok
+            store_attendees[attendee] = value
+
+        successful_stores.append(successful_stores_per_iter)
+        await asyncio.sleep(delay)
+
+        get_start = time.perf_counter()
+        get_peers = random.sample(peers, min(num_get_peers, len(peers)))
+        get_tasks = [peer.get(key, latest, return_future=True) for peer in get_peers]
+        get_result, _ = await asyncio.wait(get_tasks, return_when=asyncio.ALL_COMPLETED)
```
Comment (justheuristic) [inline benchmarks/benchmark_dht.py:69]:
```suggestion
        get_result = await asyncio.gather(*get_tasks)
```

**Review by justheuristic — CHANGES_REQUESTED** [review]

**Review by justheuristic — CHANGES_REQUESTED** [review]

File: .github/workflows/run-benchmarks.yml [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+      - name: Build hivemind
+        run: |
+          pip install .
+      - name: Benchmark
+        run: |
+          cd benchmarks
+          python benchmark_throughput.py --preset minimalistic
+          python benchmark_tensor_compression.py
+          python benchmark_dht.py
+
```
Comment (justheuristic) [inline .github/workflows/run-benchmarks.yml:36]: <\n>

**Review by justheuristic — APPROVED** [review]

**Pull request merged**